array.go

```diff
@@ -2336,22 +2670,145 @@ func NewArrayWithRootID(storage SlabStorage, rootID SlabID) (*Array, error) {
 	}
 
 	return &Array{
-		Storage: storage,
-		root:    root,
+		Storage:             storage,
+		root:                root,
+		mutableElementIndex: make(map[ValueID]uint64),
 	}, nil
 }
 
+// TODO: maybe optimize this
+func (a *Array) incrementIndexFrom(index uint64) {
+	// Although range loop over Go map is not deterministic, it is OK
+	// to use here because this operation is free of side-effect and
+	// leads to the same results independent of map order.
+	for id, i := range a.mutableElementIndex {
+		if i >= index {
+			a.mutableElementIndex[id]++
+		}
+	}
+}
+
+// TODO: maybe optimize this
+func (a *Array) decrementIndexFrom(index uint64) {
+	// Although range loop over Go map is not deterministic, it is OK
+	// to use here because this operation is free of side-effect and
+	// leads to the same results independent of map order.
+	for id, i := range a.mutableElementIndex {
+		if i > index {
+			a.mutableElementIndex[id]--
+		}
+	}
+}
+
+func (a *Array) getIndexByValueID(id ValueID) (uint64, bool) {
+	index, exist := a.mutableElementIndex[id]
+	return index, exist
+}
+
+func (a *Array) setParentUpdater(f parentUpdater) {
+	a.parentUpdater = f
+}
+
+// setCallbackWithChild sets up callback function with child value so
+// parent array a can be notified when child value is modified.
+func (a *Array) setCallbackWithChild(i uint64, child Value, maxInlineSize uint64) {
+	c, ok := child.(mutableValueNotifier)
+	if !ok {
+		return
+	}
+
+	vid := c.ValueID()
+
+	// Index i will be updated with array operations, which affects element index.
+	a.mutableElementIndex[vid] = i
+
+	c.setParentUpdater(func() error {
+
+		// Avoid unnecessary write operation on parent container.
+		// Child value was stored as SlabIDStorable (not inlined) in parent container,
+		// and continues to be stored as SlabIDStorable (still not inlinable),
+		// so no update to parent container is needed.
+		if !c.Inlined() && !c.Inlinable(maxInlineSize) {
+			return nil
+		}
+
+		// Get latest index by child value ID.
+		index, exist := a.getIndexByValueID(vid)
+		if !exist {
+			return NewFatalError(fmt.Errorf("failed to get index for child element with value id %s", vid))
+		}
+
+		// Set child value with parent array using updated index.
+		// Set() calls c.Storable() which returns inlined or not-inlined child storable.
+		existingValueStorable, err := a.Set(index, c)
+		if err != nil {
+			return err
+		}
+
+		// Verify overwritten storable has identical value ID.
+
+		switch x := existingValueStorable.(type) {
```

Comment:
Maybe just "rebind" the variable to increase readability:
```suggestion
		switch existingValueStorable := existingValueStorable.(type) {
```
and use  `existingValueStorable` instead of `x` in the cases